src/rust/engine/graph/src/entry.rs

```diff
@@ -59,29 +59,38 @@ impl Generation {
 /// same values as they did when this Node was last run; if so, the value can be re-used
 /// (and should be marked "Clean").
 ///
+/// If the value is Uncacheable it may only be consumed in the Session that produced it, and should
+/// be recomputed in a new Session.
+///
 /// If the value is Clean, the consumer can simply use the value as-is.
 ///
 #[derive(Clone, Debug)]
 pub(crate) enum EntryResult<N: Node> {
   Clean(Result<N::Item, N::Error>),
   Dirty(Result<N::Item, N::Error>),
+  Uncacheable(
+    Result<N::Item, N::Error>,
+    <<N as Node>::Context as NodeContext>::SessionId,
+  ),
 }
 
 impl<N: Node> EntryResult<N> {
-  fn is_dirty(&self) -> bool {
-    if let EntryResult::Dirty(..) = self {
-      true
-    } else {
-      false
+  fn is_clean(&self, context: &N::Context) -> bool {
+    match self {
+      EntryResult::Clean(..) => true,
+      EntryResult::Uncacheable(_, session_id) => context.session_id() == session_id,
+      EntryResult::Dirty(..) => false,
     }
   }
 
+  /// Iff the value is Clean, mark it Dirty.
```

Comment:
Soooo, possibly? dirty means: "might run again in any session", uncacheable means: "will only run again in a new session". So the gap between them is that if something uncacheable is dirtied _during_ a session, it's unclear whether to re-run it.

I think that because we're asserting that these are not side-effecting operations, re-running them (in or out of a single session) is safe. Not clear... will take a look at what it might do to the data model.

Reply:
Yeah, I suspect if we're not going to dirty them, we shouldn't allow deps; I'm imagining something like (I'm aware we don't support non-cacheable `@rule`s _now_, but we should make the framework force us to think about it if we add them):
```python
@rule(cacheable=False)
def foo():
  executable = yield Get[Snapshot](some_path_globs)
  do_sideeffecting_thing_with(executable)
```

where the side-effecting rule has a dep, someone edits the snapshotted file, we re-snapshot because that was invalidated, but we don't re-run the rule because it's side-effecting. Erroring out (ideally on the first `get` from the side-effecting rule, rather than on invalidation), rather than silently not running the rule, feels safer/less surprising.

Reply:
Just ftr: this is _not_ supposed to represent side-effecting things. Those are still allowed only at the root.

Reply:
Oh, apologies. I just realized that your example is still relevant in a `@goal_rule` (marked uncacheable), meaning that sideeffects are in fact relevant here.

But I'm not sure I understand why your example motivates 1) erroring, or 2) re-running the uncacheable rule. If anything, it feels like that motivates not dirtying it, and keeping the "exactly once per session" semantics, which would have the effect that the rule ran once to completion per session, even if things below it and above it in the graph re-ran.

Reply:
A specific (admittedly, somewhat contrived) example I can think of is:
* Some checked-in file contains a "git branch name which should be used to fetch golden test files"
* Some uncacheable rule will run `git` to get the contents of a file at that sha. It depends on reading that file, to construct the command line to run. Something like:
```python
git_branch_name = read(some_path_in_git)
golden_value = execute("git", "show", f"{git_branch_name}:golden_file")
use_golden_value()
```

As I believe the current PR stands, if this rule is marked as once-per-session, and the contents of `some_path_in_git` changes, we won't re-run this rule.

It's unclear whether we should.
I can argue that we shouldn't re-run (because in intent, "once-per-session" isn't actually an "exactly-once" guarantee, it's an "at-least once" guarantee - the point of the feature is to say "Yeah, this otherwise would be a cache hit, but we're going to pretend it wasn't because we know it reads external state").
I can argue that it should (because something regular which it depends on was invalidated, and that's how invalidation generally works).

Not allowing deps feels like a safe thing to do, because forcing people to decompose their rules to be one of "uncacehable, but no deps" or "if any of its deps change, it will rerun" seems reasonable. (Except ugh then goal_rules still need to be special).
Re-running feels like a safe thing to do, because we're saying these things _shouldn't_ be side-effecting, so it's _safe_ for us to run these things as often as we want.
Allowing deps, but not re-running, feels like it opens up a chance for stale results and unsurprising behaviour.

Reply:
> Except ugh then goal_rules still need to be special

Right... this was my realization above. And it's less about whether we want side-effecting things deeper in the graph (we definitely do not), and more about whether we want multiple types of classifications of nodes (IMO, we probably do not any time soon).

I think the way I see this is that because cacheable rules are deterministic-ish and not side-effecting, automatically retrying them to attempt to improve the observed transactionality of rules is good and healthy, and that's what we do when we dirty nodes. But, it's not strictly required... it's best effort. Given that, not making an effort for uncacheable nodes (and not attempting to differentiate them from side-effecting nodes) seems reasonable.